common/channel/src/libra_channel_test.rs

```diff
@@ -0,0 +1,82 @@
+use crate::libra_channel;
+use crate::libra_channel::MessageQueue;
+use futures::{executor::block_on, FutureExt, StreamExt};
+use std::collections::VecDeque;
+use std::thread;
+use std::time::Duration;
+
+struct TestMessageQueue {
+    queue: VecDeque<u8>,
+}
+
+impl MessageQueue for TestMessageQueue {
+    type Message = u8;
+
+    fn push(&mut self, message: Self::Message) {
+        self.queue.push_back(message);
+    }
+
+    fn pop(&mut self) -> Option<Self::Message> {
+        self.queue.pop_front()
+    }
+}
+
+#[test]
+fn test_send_recv_order() {
+    let mq = TestMessageQueue {
+        queue: VecDeque::new(),
+    };
+    let (mut sender, mut receiver) = libra_channel::new(mq);
+    sender.put(0);
+    sender.put(1);
+    sender.put(2);
+    sender.put(3);
+    let task = async move {
+        // Ensure that messages are received in order
+        assert_eq!(receiver.select_next_some().await, 0);
+        assert_eq!(receiver.select_next_some().await, 1);
+        assert_eq!(receiver.select_next_some().await, 2);
+        assert_eq!(receiver.select_next_some().await, 3);
+        // Ensures that there is no other value which is ready
+        assert_eq!(receiver.select_next_some().now_or_never(), None);
+    };
+    block_on(task);
+}
+
+#[test]
+fn test_empty() {
+    let mq = TestMessageQueue {
+        queue: VecDeque::new(),
+    };
+    let (_, mut receiver) = libra_channel::new(mq);
+    // Ensures that there is no other value which is ready
+    assert_eq!(receiver.select_next_some().now_or_never(), None);
+}
+
+#[test]
+fn test_waker() {
+    let mq = TestMessageQueue {
+        queue: VecDeque::new(),
+    };
+    let (mut sender, mut receiver) = libra_channel::new(mq);
+    // Ensures that there is no other value which is ready
+    assert_eq!(receiver.select_next_some().now_or_never(), None);
+    let join_handle = thread::spawn(move || {
+        block_on(async {
+            assert_eq!(receiver.select_next_some().await, 0);
+        });
+        block_on(async {
+            assert_eq!(receiver.select_next_some().await, 1);
+        });
+        block_on(async {
+            assert_eq!(receiver.select_next_some().await, 2);
+        });
+    });
+    thread::sleep(Duration::from_millis(100));
```

Comment:
Why are you using sleep in a test?

Reply:
What would be a better way to test this use case?

Reply:
I'm assuming that this channel is supposed to be built for async contexts (because Stream is implemented). I would recommend sticking with either a sync or an async API. If you want this to be async, then construct two sets of futures (one that sends and one that receives on the channel) and then await on the join of the two. That way you don't need to spawn another thread and you don't need any sleeps. There is a very high likelihood that these tests will be flaky since we're hoping that the thread that does the sends, sends in time.

Reply:
I might be missing something here: how can this test be flaky? `sender.put()` should always succeed and `receiver.select_next_some().await` should wait until there is a new value added and return.

Reply:
The reason for sleep is to have a simple test for the Waker API works. The receiver initially sees that there are no values and registers a waker. 100ms later in the main thread, we call a `sender.put(0)` which should call wake() and wake up the receiver task. We do this a couple of more times.

I don't think that this should be flaky if the channel is implemented correctly. If this test ever fails or has flakiness, that would expose a bug in channel implementation.

Reply:
I think your correct in that this may not be flaky, but I generally think that sleeps shouldn't ever be required to test for a particular behavior. There are plenty of different ways to test for this that don't involve sleeping. You can explicitly call the poll method and ensure that a waker was set, for example, if that's what you're trying to test for.